Creation of generation fails with Internal server error

After fixing https://github.com/camunda/zeebe/issues/14780 we experience during testing the e2e tests several internal server error.

Due to the wrong response mappings (interface of the Rest API) we can't see what the error message contains.


[This is the error we see:](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22testbench-1-x-prod%22%0Alabels.k8s-pod%2Fapp%3D%22testbench%22;cursorTimestamp=2023-11-24T11:02:23.372826663Z;startTime=2023-11-24T10:43:57.555Z;endTime=2023-11-24T12:43:57.555817Z?project=zeebe-io)


```
jakarta.ws.rs.client.ResponseProcessingException: java.io.IOException: Internal Server Error returned from URI: https://console.cloud.ultrawombat.com/external/generations/ee489414-071a-4194-a1ae-ead4183a2ae2, requestBody:{"name":"temp_W8gjazwI7R","zeebeVersion":"gcr.io/zeebe-io/zeebe:8.2.19-SNAPSHOT-stable-8.2-9d058e55","operateVersion":null}, responseBody: Cannot read properties of undefined (reading 'map')
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterResponse(ClientInvocation.java:671) ~[testbench.jar:3.0.5-SNAPSHOT]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:428) ~[testbench.jar:3.0.5-SNAPSHOT]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:134) ~[testbench.jar:3.0.5-SNAPSHOT]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103) ~[testbench.jar:3.0.5-SNAPSHOT]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:61) ~[testbench.jar:3.0.5-SNAPSHOT]
	at jdk.proxy2.$Proxy43.cloneGeneration(Unknown Source) ~[?:?]
	at io.zeebe.clustertestbench.handler.CreateGenerationInCamundaCloudHandler.cloneGeneration(CreateGenerationInCamundaCloudHandler.java:46) ~[testbench.jar:3.0.5-SNAPSHOT]
	at io.zeebe.clustertestbench.handler.CreateGenerationInCamundaCloudHandler.handle(CreateGenerationInCamundaCloudHandler.java:31) ~[testbench.jar:3.0.5-SNAPSHOT]
	at io.zeebe.clustertestbench.handler.JobHandlerWithEnrichedLogger.handle(JobHandlerWithEnrichedLogger.java:29) ~[testbench.jar:3.0.5-SNAPSHOT]
	at io.camunda.zeebe.client.impl.worker.JobRunnableFactoryImpl.executeJob(JobRunnableFactoryImpl.java:45) ~[testbench.jar:3.0.5-SNAPSHOT]
	at io.camunda.zeebe.client.impl.worker.JobRunnableFactoryImpl.lambda$create$0(JobRunnableFactoryImpl.java:40) ~[testbench.jar:3.0.5-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.io.IOException: Internal Server Error returned from URI: https://console.cloud.ultrawombat.com/external/generations/ee489414-071a-4194-a1ae-ead4183a2ae2, requestBody:{"name":"temp_W8gjazwI7R","zeebeVersion":"gcr.io/zeebe-io/zeebe:8.2.19-SNAPSHOT-stable-8.2-9d058e55","operateVersion":null}, responseBody: Cannot read properties of undefined (reading 'map')
	at io.zeebe.clustertestbench.cloud.filter.FailedResponseFilter.filter(FailedResponseFilter.java:38) ~[testbench.jar:3.0.5-SNAPSHOT]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterResponse(ClientInvocation.java:667) ~[testbench.jar:3.0.5-SNAPSHOT]
	... 16 more

```


![testbench](https://github.com/zeebe-io/zeebe-cluster-testbench/assets/2758593/cce5d82a-c426-4af7-b770-f217a9bf6ee0)

--- a/core/src/main/java/io/zeebe/clustertestbench/handler/CreateGenerationInCamundaCloudHandler.java
+++ b/core/src/main/java/io/zeebe/clustertestbench/handler/CreateGenerationInCamundaCloudHandler.java
@@ -11,6 +11,10 @@ import io.zeebe.clustertestbench.util.StringLookup;
 import org.apache.commons.lang3.RandomStringUtils;
 
 public class CreateGenerationInCamundaCloudHandler implements JobHandler {
+
+  public static final String ERROR_CODE_ILLEGAL_GENERATION = "IllegalGeneration";
+  public static final String ERROR_MSG_EXPECTED_TO_CLONE_A_GENERATION =
+      "Expected to clone a generation with upgradeableFrom reference, but generation %s has no reference set. This will cause console API to not find the respective generation.";
   private final ExternalConsoleAPIClient consoleApiClient;
 
   public CreateGenerationInCamundaCloudHandler(final ExternalConsoleAPIClient consoleApiClient) {
@@ -23,6 +27,24 @@ public class CreateGenerationInCamundaCloudHandler implements JobHandler {
 
     final var generationName = createGenerationName();
     final var templateGeneration = lookupTemplate(input.getGenerationTemplate());
+
+    if (templateGeneration.upgradeableFrom().isEmpty()) {
+      // This a shortcut until
+      // https://github.com/camunda-cloud/camunda-cloud-management-apps/issues/1731 is fixed
+      // Normally we would expect that we can clone all kind of generations but currently this is
+      // not possible.
+      // If we reference a generation without the upgradeableFrom reference (so no reference)
+      // we retrieve weird errors which are hard to debug.
+      // https://github.com/zeebe-io/zeebe-cluster-testbench/issues/945
+      client
+          .newThrowErrorCommand(job)
+          .errorCode(ERROR_CODE_ILLEGAL_GENERATION)
+          .errorMessage(
+              String.format(ERROR_MSG_EXPECTED_TO_CLONE_A_GENERATION, input.generationTemplate))
+          .send();
+
+      return;
+    }
 
     final var zeebeImage = input.getZeebeImage();
     final var operateImage = input.getOperateImage();

--- a/core/src/test/java/io/zeebe/clustertestbench/handler/CreateGenerationInCamundaCloudHandlerTest.java
+++ b/core/src/test/java/io/zeebe/clustertestbench/handler/CreateGenerationInCamundaCloudHandlerTest.java
@@ -1,6 +1,7 @@
 package io.zeebe.clustertestbench.handler;
 
 import static io.zeebe.clustertestbench.internal.cloud.StubExternalConsoleAPIClient.DEFAULT_GENERATION_NAME;
+import static io.zeebe.clustertestbench.internal.cloud.StubExternalConsoleAPIClient.GENERATION_NAME_WITHOUT_UPGRADE_FROM;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.Assertions.entry;
@@ -92,6 +93,22 @@ class CreateGenerationInCamundaCloudHandlerTest {
 
       assertThat(request.operateVersion()).isNull();
       assertThat(request.zeebeVersion()).isNull();
+    }
+
+    @Test
+    public void shouldNotCreateGenerationWhenReferencingInvalidGeneration() throws Exception {
+      // given
+      activatedJobStub.setInputVariables(
+          Map.of(FIELD_GENERATION_TEMPLATE, GENERATION_NAME_WITHOUT_UPGRADE_FROM));
+
+      // when
+      sutCreateGenerationHandler.handle(jobClientStub, activatedJobStub);
+
+      // then
+      assertThat(activatedJobStub.hasThrownError()).isTrue();
+      assertThat(activatedJobStub.getErrorMessage())
+          .contains("Expected to clone a generation with upgradeableFrom reference")
+          .contains(GENERATION_NAME_WITHOUT_UPGRADE_FROM);
     }
 
     @Test

--- a/external-console-api-client/src/main/java/io/zeebe/clustertestbench/internal/cloud/ExternalConsoleAPIClient.java
+++ b/external-console-api-client/src/main/java/io/zeebe/clustertestbench/internal/cloud/ExternalConsoleAPIClient.java
@@ -35,5 +35,9 @@ public interface ExternalConsoleAPIClient {
   record CloneGenerationRequest(String name, String zeebeVersion, String operateVersion) {}
 
   @JsonIgnoreProperties(ignoreUnknown = true)
-  record GenerationInfo(String uuid, String name, Map<String, String> versions) {}
+  record GenerationInfo(
+      String uuid,
+      String name,
+      Map<String, String> versions,
+      List<Map<String, String>> upgradeableFrom) {}
 }

--- a/external-console-api-client/src/test/java/io/zeebe/clustertestbench/internal/cloud/StubExternalConsoleAPIClient.java
+++ b/external-console-api-client/src/test/java/io/zeebe/clustertestbench/internal/cloud/StubExternalConsoleAPIClient.java
@@ -12,6 +12,10 @@ public class StubExternalConsoleAPIClient implements ExternalConsoleAPIClient {
 
   public static final String DEFAULT_GENERATION_NAME = "default-generation";
   public static final String DEFAULT_GENERATION_UUID = "1a2b3c4f56789-1234-abcd-1a2b3c4f5678";
+
+  public static final String GENERATION_NAME_WITHOUT_UPGRADE_FROM = "generationWithoutUpgradeFrom";
+  public static final String GENERATION_UUID_WITHOUT_UPGRADE_FROM =
+      "4f4b3c4f56789-1234-abcd-1a2b3c4f5678";
   public static final String DEFAULT_ZEEBE_IMAGE = "camunda/zeebe:0.23.7";
   public static final String DEFAULT_OPERATE_IMAGE = "camunda/operate:0.23.2";
 
@@ -31,9 +35,20 @@ public class StubExternalConsoleAPIClient implements ExternalConsoleAPIClient {
     versions.put(KEY_OPERATE_IMAGE, DEFAULT_OPERATE_IMAGE);
 
     final GenerationInfo generationInfo =
-        new GenerationInfo(DEFAULT_GENERATION_UUID, DEFAULT_GENERATION_NAME, versions);
-
+        new GenerationInfo(
+            DEFAULT_GENERATION_UUID,
+            DEFAULT_GENERATION_NAME,
+            versions,
+            List.of(Map.of("uuid", "fake")));
     generationInfos.add(generationInfo);
+
+    final GenerationInfo generationInfoWithoutUpgradeFrom =
+        new GenerationInfo(
+            GENERATION_UUID_WITHOUT_UPGRADE_FROM,
+            GENERATION_NAME_WITHOUT_UPGRADE_FROM,
+            versions,
+            List.of());
+    generationInfos.add(generationInfoWithoutUpgradeFrom);
   }
 
   @Override
@@ -52,7 +67,8 @@ public class StubExternalConsoleAPIClient implements ExternalConsoleAPIClient {
           new ExternalConsoleAPIClient.GenerationInfo(
               UUID.randomUUID().toString(),
               request.name(),
-              Map.of("zeebe", zeebeVersion, "operate", operateVersion));
+              Map.of("zeebe", zeebeVersion, "operate", operateVersion),
+              List.of(Map.of("uuid", "fake")));
 
       generationInfos.add(generationInfo);
       return generationInfo;

